src/client/metadataApiRetrieve.ts

```diff
@@ -313,6 +326,63 @@ export class MetadataApiRetrieve extends MetadataTransfer<MetadataApiRetrieveSta
     }
     return new ComponentSet(components, registry);
   }
+
+  // Some bundle-like components can be partially deleted in the org, then retrieved. When this
+  // happens, the deleted files need to be deleted on the file system and added to the FileResponses
+  // that are returned by `RetrieveResult.getFileResponses()` for accuracy. The component types that
+  // support this behavior are defined in the metadata registry with `"supportsPartialDelete": true`.
+  // However, not all types can be partially deleted in the org. Currently this only applies to
+  // DigitalExperienceBundle and ExperienceBundle.
+  private handlePartialDeleteMerges(retrievedComponents: SourceComponent[], tree: ZipTreeContainer): void {
+    interface PartialDeleteComp {
+      contentPath: string;
+      contentList: string[];
+    }
+    const partialDeleteComponents = new Map<string, PartialDeleteComp>();
+    const mergeWithComponents = this.components.getSourceComponents().toArray();
+
+    mergeWithComponents.forEach((comp) => {
+      if (comp.type.supportsPartialDelete && comp.content && fs.statSync(comp.content).isDirectory()) {
+        const contentList = fs.readdirSync(comp.content);
+        partialDeleteComponents.set(comp.fullName, { contentPath: comp.content, contentList });
+      }
+    });
+
+    // if no partial delete components were in the mergeWith ComponentSet, no need to continue
+    if (partialDeleteComponents.size === 0) {
+      return;
+    }
+
+    retrievedComponents.forEach((comp) => {
+      if (comp.type.supportsPartialDelete && partialDeleteComponents.has(comp.fullName)) {
+        const localComp = partialDeleteComponents.get(comp.fullName);
+        if (localComp.contentPath && tree.isDirectory(comp.content)) {
+          const remoteContentList = tree.readDirectory(comp.content);
+          let deleteLocalComp = false;
+          while (localComp.contentList?.length) {
+            const fileName = localComp.contentList.pop();
```

Comment:
I think I used a while/pop pattern here because it's fast and the contentList could mutate without side effects and iirc I started with a different implementation where I wanted the contentList to mutate.  As it is now it doesn't matter and it could switch to be foreach if that's easier to read.

I wrote a lot of comments at the top of the function to avoid a ton of inline comments but I can add something about this section, which is comparing the local component files with the retrieved component files to detect the partial delete scenario.